wp-content/themes/pub/wporg-learn-2024/src/upcoming-online-workshops/block.php

```diff
@@ -0,0 +1,82 @@
+<?php
+namespace WordPressdotorg\Theme\Learn_2024\Upcoming_Online_Workshops;
+
+use function WPOrg_Learn\Events\{get_discussion_events};
+
+add_action( 'init', __NAMESPACE__ . '\init' );
+
+/**
+ * Registers the block using the metadata loaded from the `block.json` file.
+ * Behind the scenes, it registers also all assets so they can be enqueued
+ * through the block editor in the corresponding context.
+ *
+ * @see https://developer.wordpress.org/reference/functions/register_block_type/
+ */
+function init() {
+	register_block_type(
+		dirname( dirname( __DIR__ ) ) . '/build/upcoming-online-workshops',
+		array(
+			'render_callback' => __NAMESPACE__ . '\render',
+		)
+	);
+}
+
+/**
+ * Render the block content.
+ *
+ * @param array    $attributes Block attributes.
+ * @param string   $content    Block default content.
+ * @param WP_Block $block      Block instance.
+ *
+ * @return string Returns the block markup.
+ */
+function render( $attributes, $content, $block ) {
+	$upcoming_online_workshops = get_discussion_events();
+
+	if ( empty( $upcoming_online_workshops ) || is_wp_error( $upcoming_online_workshops ) ) {
+		$content = '<!-- wp:group {"layout":{"type":"constrained"}} -->
+		<div class="wp-block-group"><!-- wp:paragraph -->
+		<p>' . __( 'No upcoming online workshops found.', 'wporg-learn' ) . '</p>
+		<!-- /wp:paragraph --></div>
+		<!-- /wp:group -->';
+	} else {
+		$content = '<!-- wp:group {"style":{"spacing":{"blockGap":"var:preset|spacing|50"}},"className":"is-style-cards-grid","layout":{"type":"grid","columnCount":"3","minimumColumnWidth":null}} --><div class="wp-block-group is-style-cards-grid">';
+
+		foreach ( $upcoming_online_workshops as $workshop ) {
+			$timestamp = strtotime( $workshop['date_utc'] ) - (int) $workshop['date_utc_offset'];
+
+			$content .= sprintf(
+				'<!-- wp:wporg/link-wrapper -->
+				<a class="wp-block-wporg-link-wrapper" href="%1$s">
+
+					<!-- wp:heading {"level":3,"style":{"spacing":{"margin":{"top":"0","bottom":"var:preset|spacing|20"}},"typography":{"lineHeight":1.6}},"fontSize":"normal"} -->
+					<h3 class="wp-block-heading has-normal-font-size" style="margin-top:0;margin-bottom:var(--wp--preset--spacing--20);line-height:1.6">%2$s</h3>
+					<!-- /wp:heading -->
+					
+					<!-- wp:paragraph {"style":{"typography":{"fontStyle":"normal","fontWeight":"700"}},"className":"is-style-short-text","fontSize":"small"} -->
+					<p class="is-style-short-text has-small-font-size" style="font-style:normal;font-weight:700">%3$s</p>
+					<!-- /wp:paragraph -->
+					
+					<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"var:preset|color|charcoal-4"}}}},"textColor":"charcoal-4","className":"is-style-short-text","fontSize":"small"} -->
+					<p class="is-style-short-text has-charcoal-4-color has-text-color has-link-color has-small-font-size" data-date-utc="%4$s"></p>
+					<!-- /wp:paragraph -->
+		
+				</a>
+				<!-- /wp:wporg/link-wrapper -->',
+				esc_html( $workshop['url'] ),
```

Comment:
Confirming that this should be esc_html and not esc_url.

Reply:
Nice catch